
---
title: "Analysis of EpiMap"
subtitle: 'Data from Kirdhar, Hoffman, et al. Nature Neuroscience, 2018'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.Date()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
vignette: >
  %\VignetteIndexEntry{Apply dream to ChIP-seq}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
---

<!--- 
# run analysis
# cd /hpc/users/hoffmg01/work/dev_dream/dream_analysis
rmarkdown::render("src/EpiMap.Rmd", output_dir='./', intermediates_dir='./')
--->


```{r initialize, cache=FALSE, echo=FALSE, message=FALSE, results='hide'}
nthreads = 40
```


```{r load.packages, echo=FALSE, message=FALSE, results='hide'}
suppressPackageStartupMessages(library(knitr))
suppressPackageStartupMessages(library(edgeR))
suppressPackageStartupMessages(library(foreach))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(variancePartition)) 
suppressPackageStartupMessages(library(GenomicRanges))  
suppressPackageStartupMessages(library(genomation)) 
suppressPackageStartupMessages(library(gridExtra)) 
suppressPackageStartupMessages(library(ggplot2)) 

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=TRUE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE,
  cache.lazy = FALSE,
  dev = c("png", "pdf"), 
  fig.width=7, fig.height=7)

options(markdown.HTML.stylesheet = 'css/custom.css')
```

```{r load.always, cache=FALSE, echo=FALSE, message=FALSE}
suppressPackageStartupMessages(library(BiocParallel))
register(SnowParam(8, "SOCK", progressbar=TRUE))
suppressPackageStartupMessages(library(synapser))
# login once and then save info
# synLogin("user.name", "password", rememberMe=TRUE)
synLogin() 
```

```{r download}
# metadata
metadata = fread( synGet('syn5691351')$path )

metadata$CellType = factor( metadata$CellType, c('NeuN-', 'NeuN+'))
metadata = metadata[HistoneMark=='H3K27ac',]
metadata$name = gsub("^HBCC_", '', metadata$Sample_ID)
metadata = data.frame(metadata)
rownames(metadata) = metadata$name 

metadata_clinical = fread( synGet('syn5691350')$path )

metadata = merge(metadata, metadata_clinical, by="Individual_ID")
rownames(metadata) = gsub("HBCC_", "", metadata$Sample_ID)

# chip-Seq counts
chipCounts = read.table( synGet('syn8078978')$path, header=TRUE, stringsAsFactors=FALSE, sep=',', row.names=1)

# peak locations
peakLocs = readBed( synGet('syn8080422')$path )
peakLocs$names = paste0("peak_", 1:length(peakLocs))

# get overlapping peaks
isect = intersect( rownames(chipCounts), peakLocs$names)
chipCounts = chipCounts[rownames(chipCounts) %in% isect,]
peakLocs = peakLocs[peakLocs$names %in% isect ]
identical(rownames(chipCounts), peakLocs$names)

# get overlapping samples
isect = intersect(colnames(chipCounts), metadata$name)
chipCounts = chipCounts[,colnames(chipCounts) %in% isect]
metadata = metadata[metadata$name %in% isect,]

# match order
idx = match(colnames(chipCounts), metadata$name)
metadata = metadata[idx,]
identical(colnames(chipCounts), metadata$name)
```

```{r process.counts}
isexpr = rowSums(cpm(chipCounts)>1) >= 0.1*ncol(chipCounts)
peakLocs2 = peakLocs[which(isexpr)]

# Standard usage of limma/voom
countObj = DGEList( chipCounts[isexpr,] )
countObj = calcNormFactors( countObj )
design = model.matrix( ~ CellType + BrainRegion, metadata)
vobj = voom( countObj, design, plot=TRUE)

identical(peakLocs2$names, rownames(vobj))
```


```{r vp2}
ctArray = c('NeuN-', 'NeuN+')
vpList = foreach( ct = ctArray ) %do% {
  idx = which(metadata$CellType == ct)
  form = ~ (1|BrainRegion) + (1|Sex) + (1|Individual_ID)
  fitExtractVarPartModel( vobj[,idx], form, metadata[idx,])
}
names(vpList) = ctArray
```

```{r vp2.plot, fig.width=15}
figList = foreach( ct = ctArray ) %do% {
  plotVarPart( sortCols( vpList[[ct]] ), main=ct ) 
}
do.call("grid.arrange", c(figList, ncol=2))           
```

```{r dream2}
fitDream = foreach( ct = ctArray ) %do% {
  # form = ~ 0 + BrainRegion + Sex + (1|Individual_ID)
  # L = getContrast( vobj[,idx], form, metadata[idx,], c('BrainRegionDLPFC', 'BrainRegionACC'))
  # idx = which(metadata$CellType == ct)
  # fit = dream( vobj[,idx], form, metadata[idx,], L)  
  form = ~ BrainRegion + Sex + (1|Individual_ID)
  idx = which(metadata$CellType == ct)
  fit = dream( vobj[,idx], form, metadata[idx,]) 
  eBayes(fit)   
}
names(fitDream) = ctArray
```



```{r dupCor, echo=FALSE, message=FALSE}
fitDupCorList = foreach( ct = ctArray ) %do% {

  # L = getContrast( vobj[,idx], form, metadata[idx,], c('BrainRegionDLPFC', 'BrainRegionACC'))
  idx = which(metadata$CellType == ct)

  design = model.matrix( ~ BrainRegion + Sex, metadata[idx,])
   
  # Estimate linear mixed model with a single variance component
  # Fit the model for each gene, 
  dupcor <- duplicateCorrelation(vobj[,idx],design,block=metadata$Individual_ID[idx])
    
  # But this step uses only the genome-wide average for the random effect
  fitDupCor <- lmFit(vobj[,idx], design, block=metadata$Individual_ID[idx], correlation=dupcor$consensus)

  # fitDupCor = contrasts.fit( fitDupCor, L)
   
  # Fit Empirical Bayes for moderated t-statistics
  eBayes( fitDupCor )
}
names(fitDupCorList) = ctArray
```

# Compare -log10 p from dream and duplicateCorrelation
```{r compare, fig.width=15}
figCompare = foreach( ct = ctArray ) %do% {

  p1 = topTable(fitDupCorList[[ct]],number=Inf, sort.by="none", coef='BrainRegionDLPFC')$P.Value
  p2 = topTable(fitDream[[ct]], number=Inf, sort.by="none", coef='BrainRegionDLPFC')$P.Value
                    
  plotCompareP( p1, p2, with(vpList[[ct]] , Individual_ID+Sex), dupcor$consensus) + ggtitle(ct)
}
do.call("grid.arrange", c(figCompare, ncol=2))     
```

```{r}
# system("cat EpiMap.html | grep -v 'Found more than one class' | grep -v PythonEmbedInR > EpiMap2.html ")
```


