
---
title: "Analysis of EpiMap"
subtitle: 'Data from Kirdhar, Hoffman, et al. Nature Neuroscience, 2018'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.Date()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
vignette: >
  %\VignetteIndexEntry{Apply dream to ChIP-seq}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
---

<!--- 
# run analysis
# cd /hpc/users/hoffmg01/work/dev_dream/dream_analysis
rmarkdown::render("src/EpiMap.Rmd", output_dir='./', intermediates_dir='./')
--->


```{r initialize, cache=FALSE, echo=FALSE, message=FALSE, results='hide'}
nthreads = 40
```


```{r load.packages, echo=FALSE, message=FALSE, results='hide'}
suppressPackageStartupMessages(library(knitr))
suppressPackageStartupMessages(library(edgeR))
suppressPackageStartupMessages(library(foreach))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(variancePartition)) 
suppressPackageStartupMessages(library(GenomicRanges))  
suppressPackageStartupMessages(library(genomation)) 
suppressPackageStartupMessages(library(gridExtra)) 
suppressPackageStartupMessages(library(ggplot2)) 

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=TRUE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE,
  cache.lazy = FALSE,
  dev = c("png", "pdf"), 
  fig.width=7, fig.height=7)

options(markdown.HTML.stylesheet = 'css/custom.css')
```

```{r load.always, cache=FALSE, echo=FALSE, message=FALSE}
suppressPackageStartupMessages(library(doParallel))
if( ! exists("cl") ){
  cl <- parallel::makeForkCluster(nthreads)
  registerDoParallel(cl)
}
suppressPackageStartupMessages(library(synapser))
# login once and then save info
# synLogin("user.name", "password", rememberMe=TRUE)
synLogin() 
```

```{r download}
# metadata
metadata = fread( synGet('syn5691351')$path )

metadata$CellType = factor( metadata$CellType, c('NeuN-', 'NeuN+'))
metadata = metadata[HistoneMark=='H3K27ac',]
metadata$name = gsub("^HBCC_", '', metadata$Sample_ID)
metadata = data.frame(metadata)
rownames(metadata) = metadata$name 

metadata_clinical = fread( synGet('syn5691350')$path )

metadata = merge(metadata, metadata_clinical, by="Individual_ID")

# chip-Seq counts
chipCounts = read.table( synGet('syn8078978')$path, header=TRUE, stringsAsFactors=FALSE, sep=',', row.names=1)

# peak locations
peakLocs = readBed( synGet('syn8080422')$path )
peakLocs$names = paste0("peak_", 1:length(peakLocs))

# get overlapping peaks
isect = intersect( rownames(chipCounts), peakLocs$names)
chipCounts = chipCounts[rownames(chipCounts) %in% isect,]
peakLocs = peakLocs[peakLocs$names %in% isect ]
identical(rownames(chipCounts), peakLocs$names)

# get overlapping samples
isect = intersect(colnames(chipCounts), metadata$name)
chipCounts = chipCounts[,colnames(chipCounts) %in% isect]
metadata = metadata[metadata$name %in% isect,]

# match order
idx = match(colnames(chipCounts), metadata$name)
metadata = metadata[idx,]
identical(colnames(chipCounts), metadata$name)
```

```{r process.counts}
isexpr = rowSums(cpm(chipCounts)>1) >= 0.1*ncol(chipCounts)
peakLocs2 = peakLocs[which(isexpr)]

# Standard usage of limma/voom
countObj = DGEList( chipCounts[isexpr,] )
countObj = calcNormFactors( countObj )
design = model.matrix( ~ CellType + BrainRegion, metadata)
vobj = voom( countObj, design, plot=TRUE)

identical(peakLocs2$names, rownames(vobj))
```


```{r vp}
form = ~ (1|CellType) + (1|BrainRegion) + (1|Sex) + (1|Individual_ID)
vp = fitExtractVarPartModel( vobj, form, metadata)
```

```{r vp.plot}
plotVarPart( sortCols( vp ) )           
```


```{r vp2}
idx = which(metadata$CellType == 'NeuN+')
form = ~ (1|BrainRegion) + (1|Sex) + (1|Individual_ID)
vp2 = fitExtractVarPartModel( vobj[,idx], form, metadata[idx,])
```

```{r vp2.plot}
plotVarPart( sortCols( vp2 ) )           
```

```{r dream}
form = ~ 0 + CellType:BrainRegion + (1|Individual_ID)

# Compare NeuN in DLPFC
L1 = getContrast( vobj, form, metadata, c('CellTypeNeuN-:BrainRegionDLPFC', 'CellTypeNeuN+:BrainRegionDLPFC'))

# Compare NeuN in DLPFC
L2 = getContrast( vobj, form, metadata, c('CellTypeNeuN-:BrainRegionACC', 'CellTypeNeuN+:BrainRegionACC'))

# Compare brain region in NeuN+
L3 = getContrast( vobj, form, metadata, c('CellTypeNeuN+:BrainRegionDLPFC', 'CellTypeNeuN+:BrainRegionACC'))

# Compare brain region in NeuN-
L4 = getContrast( vobj, form, metadata, c('CellTypeNeuN-:BrainRegionDLPFC', 'CellTypeNeuN-:BrainRegionACC'))
  
L = cbind(L1, L2, L3, L4)

```
```{r dream2}
fit = dream( vobj, form, metadata, L)
fit = eBayes(fit)   
```


```{r dupCor, echo=FALSE, message=FALSE}
design = model.matrix( ~ 0 + CellType:BrainRegion, metadata)
 
# Estimate linear mixed model with a single variance component
# Fit the model for each gene, 
dupcor <- duplicateCorrelation(vobj,design,block=metadata$Individual_ID)

# But this step uses only the genome-wide average for the random effect
fitDupCor <- lmFit(vobj, design, block=metadata$Individual_ID, correlation=dupcor$consensus)

fitDupCor = contrasts.fit( fitDupCor, L)
 
# Fit Empirical Bayes for moderated t-statistics
fitDupCor <- eBayes( fitDupCor )
```


# Compare -log10 p from dream and duplicateCorrelation
```{r compare}
figList = foreach( i = 1:4) %do% {
  id = colnames(coef(fit))[i] 
  p1 = topTable(fitDupCor, coef=id,number=Inf, sort.by="none")$P.Value
  p2 = topTable(fit, coef=id, number=Inf, sort.by="none")$P.Value
         
  plotCompareP( p1, p2, with(vp2, Individual_ID+Sex), dupcor$consensus)
}
do.call("grid.arrange", c(figList, ncol=2))
```


# show DE results
```{r de}
# df = topTable(fit, number=Inf)
 
# hist( df$t ) 
# hist( df$logFC )
 
# ggplot(df, aes(logFC, -log10(P.Value)) ) + geom_point() + theme_bw(12)
```
